aptos-move/block-executor/src/code_cache_global.rs

```diff
@@ -64,88 +53,90 @@ where
         self.valid.load(Ordering::Acquire)
     }
 
-    /// Returns the module code stored is this [ImmutableModuleCode].
-    fn inner(&self) -> &Arc<ModuleCode<DC, VC, E, Option<TxnIndex>>> {
-        self.module.deref()
+    /// Returns the module code stored is this [Entry].
+    fn module_code(&self) -> &Arc<ModuleCode<Deserialized, Verified, Extension>> {
+        &self.module
     }
 }
 
-/// An immutable cache for verified code, that can be accessed concurrently thought the block, and
-/// only modified at block boundaries.
-pub struct ImmutableModuleCache<K, DC, VC, E> {
+/// A global module cache for verified code that is read-only and concurrently accessed during the
+/// block execution. Modified safely only at block boundaries.
+pub struct GlobalModuleCache<K, D, V, E> {
```

Comment:
I would have `BlockExecutor` take a reference to `RWLockReadGuard<'_. GlobalModuleCache>` or something that wraps it. The overhead is negligible compared to block execution time and everything else, and it's much better design, and clearly describes the intention.

On block boundaries we acquire the Write guard and operate on it.

Reply:
I think instead of ExplicitSyncWrapper on the members, basically we can pass a GlobalModuleCache in a guard. But I think instead of RWLock or Mutex guard, we can have a guard that asserts that concurrent blocks don't use the same thing - but theoretically, if two blocks have the same parent, it could make sense to use the cache in the same state, so RWGuard wrapped GlobalModuleCache without ExplicitSyncWrappers inside seems to work well.

Reply:
in fact using (rw) lock guard is perfect - no degradation for acquiring the locks, but forcing the block execution itself to have acquired exclusive access.

Reply:
Indeed, would be beautiful, thanks @ziaptos @gelash! Let me try to do that (maybe even in this PR)